### PR TITLE
add internal/cache/gache test

### PR DESCRIPTION
**Pull request by vankichi**
Signed-off-by: vankichi <kyukawa315@gmail.com>

### Description:

I added test of `internal/cache/gache/gache.go` and updated function comment about `internal/cache/gache/gache.go`.
Also, a bit fix in `internal/cache/gache/option_test.go`.

### Environment:

- Go Version: 1.14.4
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.0

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/cache/gache/gache_test.go [hunk 1]
```diff
@@ -397,69 +379,62 @@ func Test_cache_Delete(t *testing.T) {
...
+			args: args{
+				key: "vdaas",
+			},
+			fields: fields{
+				gache:          gache.New(),
+				expireDur:      1 * time.Second,
+				expireCheckDur: 1 * time.Second,
+				expiredHook:    nil,
+			},
+			checkFunc: defaultCheckFunc,
```
Comment (hlts2) [inline internal/cache/gache/gache_test.go:400]: Could you please check if the object was dleted in to the `fields.gache`?

File: internal/cache/gache/gache_test.go [hunk 2]
```diff
@@ -312,49 +318,25 @@ func Test_cache_Set(t *testing.T) {
...
+				key: "vdaas",
+				val: "vald",
+			},
+			fields: fields{
+				gache:          gache.New(),
+				expireDur:      1 * time.Second,
+				expireCheckDur: 1 * time.Second,
+				expiredHook:    nil,
+			},
+			checkFunc: defaultCheckFunc,
```
Comment (hlts2) [inline internal/cache/gache/gache_test.go:333]: Could you please check if the object was created in to the `fields.gache`?

File: internal/cache/gache/gache_test.go (hunk 1 above)
Comment (vankichi) [reply internal/cache/gache/gache_test.go:400]: @hlts2  sorry, I could not understand what you mentioned 😢 
Could you teach me more detail?

Comment (hlts2) [reply internal/cache/gache/gache_test.go:400]: ok!
For example, if we call `Delete` method, `Delete` method of `c.gache.Delete` is called.
So we need to check the objects in the gash are gone.

Comment (kevindiu) [reply internal/cache/gache/gache_test.go:400]: is it the responsibility of `gache` library? (same as `file/watcher` we discussed before)

File: internal/cache/gache/gache_test.go [hunk 3]
```diff
@@ -43,41 +52,61 @@ func TestNew(t *testing.T) {
 		afterFunc  func(args)
 	}
 	defaultCheckFunc := func(w want, gotC *cache) error {
-		if !reflect.DeepEqual(gotC, w.wantC) {
+		opts := []cmp.Option{
+			cmp.AllowUnexported(*w.wantC),
+			cmp.AllowUnexported(*gotC),
+			cmp.Comparer(func(want, got *cache) bool {
+				return want.gache != nil && got.gache != nil
```
Comment (hlts2) [inline internal/cache/gache/gache_test.go:59]:
```
opts := []cmp.Option{
	cmp.AllowUnexported(*w.wantC),
	cmp.AllowUnexported(*gotC),
	cmp.Comparer(func(want, got gache.Gache) bool {
		return want != nil && got != nil
	}),
}
if diff := cmp.Diff(w.wantC, gotC, opts...); diff != "" {
	return errors.Errorf("got = %v, want %v", gotC, w.wantC)
}
return nil
```
before: cmp.Comparer(func(want, got *cache) bool {
after: cmp.Comparer(func(want, got gache.Gache) bool {
↑ You can check only gache field with  custom compare function.

File: internal/cache/gache/gache_test.go [hunk 4]
```diff
@@ -43,41 +52,61 @@ func TestNew(t *testing.T) {
 		afterFunc  func(args)
 	}
 	defaultCheckFunc := func(w want, gotC *cache) error {
-		if !reflect.DeepEqual(gotC, w.wantC) {
+		opts := []cmp.Option{
+			cmp.AllowUnexported(*w.wantC),
+			cmp.AllowUnexported(*gotC),
+			cmp.Comparer(func(want, got *cache) bool {
```
Comment (hlts2) [inline internal/cache/gache/gache_test.go:58]:
```suggestion
			cmp.Comparer(func(want, got gache.Gache) bool {
```

File: internal/cache/gache/gache_test.go [hunk 5]
```diff
@@ -43,41 +52,64 @@ func TestNew(t *testing.T) {
...
 	}
 	defaultCheckFunc := func(w want, gotC *cache) error {
-		if !reflect.DeepEqual(gotC, w.wantC) {
-			return errors.Errorf("got = %v, want %v", gotC, w.wantC)
+		opts := []cmp.Option{
+			cmp.AllowUnexported(*w.wantC),
+			cmp.AllowUnexported(*gotC),
+			cmp.Comparer(func(want, got gache.Gache) bool {
+				return want != nil && got != nil
+			}),
```
Comment (hlts2) [inline internal/cache/gache/gache_test.go:60]:
```suggestion
			}),
			cmp.Comparer(func(want, got func(context.Context, string)) bool
				return reflect.ValueOf(want).Pointer() == reflect.ValueOf(want).Pointer()
			}
```

File: internal/cache/gache/gache_test.go [hunk 6]
```diff
@@ -43,41 +52,64 @@ func TestNew(t *testing.T) {
...
+		}
+		if diff := cmp.Diff(w.wantC, gotC, opts...); diff != "" {
+			if reflect.ValueOf(w.wantC.expiredHook).Pointer() != reflect.ValueOf(gotC.expiredHook).Pointer() {
+				return errors.Errorf("got = %v, want %v", gotC, w.wantC)
+
+			}
```
Comment (hlts2) [inline internal/cache/gache/gache_test.go:66]:
```suggestion
			
```

File: internal/cache/gache/gache_test.go [hunk 7]
```diff
@@ -502,62 +515,63 @@ func Test_cache_GetAndDelete(t *testing.T) {
...
+				expiredHook:    nil,
+			},
+			want: want{
+				want:  "vald",
+				want1: true,
+			},
+			beforeFunc: func(args args, c *cache) {
+				c.Set(args.key, "vald")
+			},
+			checkFunc: defaultCheckFunc,
```
Comment (kevindiu) [inline internal/cache/gache/gache_test.go:553]:
```suggestion
```

File: internal/cache/gache/gache_test.go [hunk 8]
```diff
@@ -390,79 +390,92 @@ func Test_cache_Delete(t *testing.T) {
...
+			},
+			want: want{
+				key:   "vdaas",
+				want:  nil,
+				want1: false,
+			},
+			beforeFunc: func(args args, c *cache) {
+				c.Set(args.key, "vald")
+			},
+			checkFunc: defaultCheckFunc,
```
Comment (kevindiu) [inline internal/cache/gache/gache_test.go:454]:
```suggestion
```

File: internal/cache/gache/gache_test.go [hunk 9]
```diff
@@ -390,79 +390,92 @@ func Test_cache_Delete(t *testing.T) {
...
+				expireDur:      1 * time.Second,
+				expireCheckDur: 1 * time.Second,
+				expiredHook:    nil,
+			},
+			want: want{
+				key:   "vdaas",
+				want:  nil,
+				want1: false,
+			},
+			checkFunc: defaultCheckFunc,
```
Comment (kevindiu) [inline internal/cache/gache/gache_test.go:433]:
```suggestion
```

File: internal/cache/gache/gache_test.go [hunk 10]
```diff
@@ -298,63 +307,54 @@ func Test_cache_Set(t *testing.T) {
...
+				expireDur:      1 * time.Second,
+				expireCheckDur: 1 * time.Second,
+				expiredHook:    nil,
+			},
+			want: want{
+				key:   "vdaas",
+				want:  "vald",
+				want1: true,
+			},
+			checkFunc: defaultCheckFunc,
```
Comment (kevindiu) [inline internal/cache/gache/gache_test.go:351]:
```suggestion
```

File: internal/cache/gache/gache_test.go [hunk 11]
```diff
@@ -220,68 +230,67 @@ func Test_cache_Get(t *testing.T) {
...
+				expiredHook:    nil,
+			},
+			want: want{
+				want:  "vald",
+				want1: true,
+			},
+			beforeFunc: func(args args, c *cache) {
+				c.Set(args.key, "vald")
+			},
+			checkFunc: defaultCheckFunc,
```
Comment (kevindiu) [inline internal/cache/gache/gache_test.go:268]:
```suggestion
```

File: internal/cache/gache/gache_test.go [hunk 12]
```diff
@@ -220,68 +230,67 @@ func Test_cache_Get(t *testing.T) {
...
+				gache:          gache.New(),
+				expireDur:      1 * time.Second,
+				expireCheckDur: 1 * time.Second,
+				expiredHook:    nil,
+			},
+			want: want{
+				want:  nil,
+				want1: false,
+			},
+			checkFunc: defaultCheckFunc,
```
Comment (kevindiu) [inline internal/cache/gache/gache_test.go:248]:
```suggestion
```

File: internal/cache/gache/gache_test.go [hunk 13]
```diff
@@ -122,47 +152,28 @@ func Test_cache_Start(t *testing.T) {
...
+					return ctx
+				}(),
+			},
+			fields: fields{
+				gache:          gache.New(),
+				expireDur:      1 * time.Second,
+				expireCheckDur: 1 * time.Second,
+				expiredHook:    nil,
+			},
+			checkFunc: defaultCheckFunc,
```
Comment (kevindiu) [inline internal/cache/gache/gache_test.go:170]:
```suggestion
```

File: internal/cache/gache/gache_test.go [hunk 14]
```diff
@@ -43,41 +52,64 @@ func TestNew(t *testing.T) {
...
+				name: "set success when opts is not nil",
+				args: args{
+					opts: []Option{
+						WithExpiredHook(expiredHook),
+					},
+				},
+				want: want{
+					wantC: c,
+				},
+				checkFunc: defaultCheckFunc,
```
Comment (kevindiu) [inline internal/cache/gache/gache_test.go:105]:
```suggestion
```

File: internal/cache/gache/gache_test.go [hunk 15]
```diff
@@ -43,41 +52,64 @@ func TestNew(t *testing.T) {
...
+			for _, opt := range defaultOptions() {
+				opt(c)
+			}
+			c.gache.SetDefaultExpire(c.expireDur)
+			return test{
+				name: "set success when opts is nil",
+				want: want{
+					wantC: c,
+				},
+				checkFunc: defaultCheckFunc,
```
Comment (kevindiu) [inline internal/cache/gache/gache_test.go:82]:
```suggestion
```

**Pull request merged**